libkbfs/merkle_checker.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"bytes"
+	"context"
+
+	merkle "github.com/keybase/go-merkle-tree"
+	"github.com/keybase/kbfs/kbfsmd"
+	"github.com/keybase/kbfs/tlf"
+	"github.com/pkg/errors"
+)
+
+// merkleChecker implements a storage engine that can be used to
+// verify a chain of merkle nodes returned from the server, to
+// validate a given MD leaf.
+type merkleChecker struct {
+	root     *kbfsmd.MerkleRoot
+	nodes    [][]byte
+	nextNode int
+}
+
+var _ merkle.StorageEngine = (*merkleChecker)(nil)
+
+func (mc *merkleChecker) StoreNode(
+	_ context.Context, _ merkle.Hash, _ []byte) error {
+	return errors.New("merkleChecker can't store nodes")
+}
+
+func (mc *merkleChecker) CommitRoot(
+	_ context.Context, _ merkle.Hash, _ merkle.Hash, _ merkle.TxInfo) error {
+	return errors.New("merkleChecker can't commit roots")
+}
+
+func (mc *merkleChecker) LookupNode(
+	_ context.Context, h merkle.Hash) ([]byte, error) {
+	// Assume the hash refers to the next node in the list; if not,
+	// the tree will fail when verifying.
+	if mc.nextNode >= len(mc.nodes) {
+		return nil, errors.Errorf(
+			"Can't lookup node %d when there are only %d nodes",
+			mc.nextNode, len(mc.nodes))
+	}
+	n := mc.nodes[mc.nextNode]
+	mc.nextNode++
+	return n, nil
+}
+
+func (mc *merkleChecker) LookupRoot(_ context.Context) (
+	h merkle.Hash, err error) {
+	return mc.root.Hash, nil
+}
+
+func verifyMerkleNodes(
+	ctx context.Context, kbfsRoot *kbfsmd.MerkleRoot, nodes [][]byte,
+	tlfID tlf.ID) error {
+	// Verify the merkle nodes by pretending to look up the nodes
+	// using a merkle.Tree, which verifies all the nodes along the
+	// path of the lookup.
+	mc := &merkleChecker{kbfsRoot, nodes, 0}
+	config := merkle.NewConfig(
+		merkle.SHA512Hasher{}, 256, 512, kbfsmd.MerkleLeaf{})
+	tree := merkle.NewTree(mc, config)
+	foundLeaf, rootHash, err := tree.Find(ctx, merkle.Hash(tlfID.Bytes()))
```

Comment:
Maybe add a comment here to mention this is where it's gonna error if hash of node returned by `LookupNode` doesn't match what we expect?